// @tact-lang/emulator/dist/utils/createAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAccount = void 0;
const ton_core_1 = require("ton-core");
function createAccount(args) {
    let workchain;
    if (args.workchain !== undefined && args.workchain !== null) {
        workchain = args.workchain;
    }
    else {
        workchain = 0;
    }
    let address;
    if (args.address) {
        address = args.address;
    }
    else {
        address = (0, ton_core_1.contractAddress)(workchain, { code: args.code, data: args.data });
    }
    let balance;
    if (args.balance) {
        balance = args.balance;
    }
    else {
        balance = 0n;
    }
    return {
        addr: address,
        storageStats: {
            used: { cells: 0n, bits: 0n, publicCells: 0n },
            lastPaid: 0
        },
        storage: {
            lastTransLt: 0n,
            balance: { coins: balance },
            state: {
                type: 'active',
                state: {
                    code: args.code,
                    data: args.data
                }
            }
        }
    };
}
exports.createAccount = createAccount;
